File: Deploy/RapierLoom.Examples/03-SimpleTracingInstead/VB/TracingAspect.vb
Public Class TracingAspect
    Inherits AspectAttribute

    <[Call](Advice.Before), Include("SayHello")> _
    Public Sub Foo(ByVal name As String)
        Console.WriteLine("This is before entering a method.")
    End Sub

    <IncludeAll(), [Call](Advice.Before)> _
    Public Sub Foo2(<JPRetVal()> ByVal retval As String, ByVal name As String)
        Console.WriteLine("This is before entering a method with the string parameter '{0}' which returns a string value.", name)
    End Sub

    <[Call](Advice.Before), IncludeAll()> _
    Public Sub Foo3(<JPRetVal()> ByVal retval As String, ByVal param1 As String, ByVal param2 As String)
        Console.WriteLine("This is before entering a method with two string parameters '{0}', '{1}' which returns a string value.", param1, param2)
    End Sub

    <[Call](Advice.Around)> _
    Public Sub SayHello(ByVal name As String)
        Console.WriteLine("This is not the SayHello method. But I know who you are {0}.", name)
    End Sub

End Class

File: Deploy/RapierLoom.Examples/02-SimpleTracingBeforeAndAfter/CS/TracingAspect.cs
// ----------------------------------------------------------------
// * The LOOM.NET project http://rapier-loom.net *
// (C) Copyright by Wolfgang Schult. All rights reserved. 
// This code is licensed under the Apache License 2.0
// To get more information, go to http://loom.codeplex.com/license
// ----------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Text;
using Loom.JoinPoints;
using Loom;

namespace SimpleTracingBeforeAndAfter
{
    /// <summary>
    /// This is a simple tracing aspect. Every aspect has to inherit the AspectAttribute.
    /// </summary>
    public class TracingAspect : AspectAttribute
    {
        /// <summary>
        /// This aspect will be interwoven with methods matching this signature.
        ///  return value doesn't matter; use [JPRetval] attribute to define specific return type
        ///  method name is exactly "SayHello"
        ///  one method parameter of type string and any name
        /// </summary>
        /// <param name="name">the name of the parameter doesn't matter</param>
        [Call(Advice.Before)] // JoinPoint, which indicates that this aspect is called before the target method 
        public void SayHello(string name) 
        {
            Console.WriteLine("This is before entering method SayHello.");
        }

        /// <summary>
        /// This aspect will be interwoven with methods matching this signature.
        ///  return value doesn't matter; use [JPRetval] attribute to define specific return type
        ///  method name is exactly "SayHello", not Foo - see Include attribute
        ///  one method parameter of type string and any name
        /// </summary>
        /// <param name="name"></param>
        [Call(Advice.After)] // JoinPoint, which indicates that this aspect is called after the target method 
        [Include("SayHello")]  // includes any method with the following signature and the method name SayHello      
        public void Foo(string name) // the name Foo of the aspect method doesn't matter now
        {
            Console.WriteLine("This is after leaving method SayHello.");
        }

    }
}
